caproto/threading/client.py

```diff
@@ -242,29 +237,27 @@ def remove_socket(self, sock):
                 #              '(obj = %s)', sock, obj)
                 self._unregister_sockets.add(sock)
 
-    def _object_removed(self, obj_id):
-        with self._socket_map_lock:
-            if obj_id in self.id_to_socket:
-                sock = self.id_to_socket.pop(obj_id)
-                # self.log.debug('Object ID %s was destroyed: removing %s', obj_id,
-                #              sock)
-                del self.socket_to_id[sock]
-                self._unregister_sockets.add(sock)
-
     def __call__(self):
         '''Selector poll loop'''
         avail_buf = array.array('i', [0])
-        while not self._close_event.is_set():
+        while self.running:
             with self._socket_map_lock:
                 for sock in self._unregister_sockets:
                     self.selector.unregister(sock)
+                self._socket_count -= len(self._unregister_sockets)
                 self._unregister_sockets.clear()
 
                 for sock, obj_id in self._register_sockets.items():
                     self.selector.register(sock, selectors.EVENT_READ,
                                            data=obj_id)
+                self._socket_count += len(self._register_sockets)
                 self._register_sockets.clear()
 
+            if self._socket_count == 0:
+                if self._register_event.wait(timeout=0.1):
+                    self._register_event.clear()
```

Comment:
When trio removed the `clear()` method from `Event` I took the lesson to heart that, “Once an Event happens it cannot un-happen,” and that in general it’s better/safer to just make a new `Event`. That said, the alternative here is to put a boolean inside a lock, which seems silly if we can just use `Event` exactly as you have done.

Reply:
Yeah, I fiddled around with this for a bit and thought it to be the lesser evil.